packages/fast-tooling-react/src/navigation/navigation-tree-item.tsx

```diff
@@ -1,255 +1,216 @@
-        return {
-            dataLocation: props.dataLocation,
-        };
-    },
-    endDrag: (props: NavigationTreeItemProps): void => {
-        props.onDragHover(null);
-    },
-};
-
-export const navigationTreeItemDropSource: DropTargetSpec<NavigationTreeItemProps> = {
-    drop: (
-        props: NavigationTreeItemProps,
-        monitor: DropTargetMonitor,
-        component: any
-    ): void => {
-        if (monitor.didDrop()) {
-            return;
-        }
-
-        const item: any = monitor.getItem();
-
-        if (props.dataLocation === item.dataLocation) {
-            return;
-        }
-
-        if (
-            props.type === NavigationDataType.children ||
-            props.type === NavigationDataType.component ||
-            (props.type === NavigationDataType.primitiveChild &&
-                (props.dragHoverAfter || props.dragHoverBefore))
-        ) {
-            const verticalDirection: VerticalDragDirection = props.dragHoverAfter
-                ? VerticalDragDirection.down
-                : props.dragHoverBefore
-                    ? VerticalDragDirection.up
-                    : VerticalDragDirection.center;
-
-            component.props.onChange(
-                item.dataLocation,
-                props.dataLocation,
-                props.type,
-                verticalDirection
-            );
-        }
-    },
-    hover: (
-        props: NavigationTreeItemProps,
-        monitor: DropTargetMonitor,
-        component: any
-    ): void => {
-        if (
-            monitor.isOver({ shallow: true }) &&
-            props.type === NavigationDataType.children
-        ) {
-            props.onDragHover(props.dataLocation);
-        } else if (
-            monitor.isOver({ shallow: true }) &&
-            (props.type === NavigationDataType.component ||
-                props.type === NavigationDataType.primitiveChild)
-        ) {
-            if (!component) {
-                return null;
-            }
-
-            const node: HTMLDivElement = component.getNode();
-
-            if (!node) {
-                return null;
-            }
-
-            const hoverBoundingRect: ClientRect | DOMRect = node.getBoundingClientRect();
-            const hoverMiddleY: number =
-                (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2; // vertical centerline
-            const clientOffset: XYCoord | null = monitor.getClientOffset(); // mouse position
-            const hoverClientY: number = clientOffset.y - hoverBoundingRect.top; // pixels from the top
-            const verticalOffset: number = hoverBoundingRect.height / 4;
-            const direction: VerticalDragDirection =
-                hoverClientY < hoverMiddleY - verticalOffset
-                    ? VerticalDragDirection.up
-                    : hoverClientY > hoverMiddleY + verticalOffset
-                        ? VerticalDragDirection.down
-                        : VerticalDragDirection.center;
+function treeItemEndLeaf(
```

Comment:
Any reason you're not making these functional components? I think you could simplify all of these function arguments into a prop interface and use this like a component, which will be more extensible over time and is more idiomatic:

```
<TreeItemEndLeaf
  ...props
/>
```

Reply:
Because, oddly, once you do that the drag and drop functionality ceases to work. I haven't nailed down why that would be but it seems that drag and drop is expecting a single component to work with.